### PR TITLE
Remove MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3 option

### LINKED ISSUE #4386: Remove support for extensions in pre-v3 X.509 certificates
### Context

X.509 is the standard for certificates used, among other things, in TLS. The original standard (v1) was published by the ITU in 1988, followed by v2 in 1993 and v3 in 1997. The IETF defines a "Internet X.509 Public Key Infrastructure" profile of X.509, based on v3 since its inception in 2002 (RFC 3280). The CA/Browser forum defines baseline requirements for use on the web, also based on v3 since their inception in 2011.

Version 3 of X.509 introduced the possibility for certificates to include optional extensions. This possibility was not present in earlier versions of the specification, however some software did generate pre-v3 certificates that did include extensions. Such certificates are not and never were compliant, so by default Mbed TLS rejects them. 

However, in order to support users faced with such certificates (generated by 3rd parties) a compile-time option `MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3` was introduced in 2013, that would cause such certificates to be accepted if they are otherwise compliant and valid.

### Proposal

Remove the option `MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3`  in Mbed TLS 3.0 and just always reject non-compliant certificates that declare a version lower than 3 but include extensions.

### Rationale

- This is 2021 and hopefully all certificates that were generated recently enough to not be expired yet are v3, so we shouldn't need to care about pre-v3 certs, let alone non-compliant pre-v3 certificates.
- Every additional compile-time option brings some complexity that negatively impacts maintainability and testing. This is only OK if this cost is justified by the service the option provides to users, but that doesn't seem to be the case here.

### Work to do for 3.0

- Remove `MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3` and let the code behave as if it were always disabled.
- Remove tests that depend on this option; for tests that depend on its negation, remove the dependency declaration.
- As a result, `git grep MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3` should find no occurrence (except in the ChangeLog) and the CI should still pass.

### Mailing-list discussion

https://lists.trustedfirmware.org/pipermail/mbed-tls/2021-April/000340.html

### PR DESCRIPTION
**Pull request by TRodziewicz**
Signed-off-by: TRodziewicz <tomasz.rodziewicz@mobica.com>

## Description
Remove the option MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3 in Mbed TLS 3.0 and just always reject non-compliant certificates that declare a version lower than 3 but include extensions.

## Status
READY

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated

Fixes #4386

### CODE REVIEW

**Review by d3zd3z — APPROVED** [review]
Two minor questions that don't really affect the PR itself.
1. Do we want to test that earlier certificates are rejected?
2. This change makes the version check permanent. Is there any other code that we can remove that is there to support features of the earlier certificates?

Otherwise, looks fine

File: docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Remove the `MBEDTLS_X509_ALLOW_EXTENSIONS_NON_V3` option
+--
+
+This change does not affect users who are working with current V3 X.509
```
Comment (mpg) [inline docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md:4]: Suggested addition: "This change does not affect people who were using the default configuration, as this option was already disabled by default. Also, it does not affect ..."

File: docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+certificates.
+
+This change makes the pre-V3 X.509 certificates both with or without optional
+extensions obsolete.
```
Comment (mpg) [inline docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md:8]: I don't think that's correct: the code change I see is only about extension, I don't think we're currently rejecting pre-v3 certificates in general. Suggested text: "Extensions were added in v3 of the X.509 specification, so pre-v3 certificates containing extensions were never compliant. Mbed TLS now rejects them with a parsing error in all configurations, as it did previously in the default configuration."

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good, but I think the migration guide entry can be improved.

File: docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md (hunk 2 above)
Comment (TRodziewicz) [reply docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md]: Done

File: docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md (hunk 1 above)
Comment (TRodziewicz) [reply docs/3.0-migration-guide.d/remove_supp_for_extensions_in_pre-v3_X_509_certs.md]: Done

**Review by mpg — APPROVED** [review]
LGTM!

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM

**Pull request merged**